crates/ark/src/treesitter.rs

```diff
@@ -404,12 +387,45 @@ impl NodeTypeExt for Node<'_> {
     fn is_binary_operator(&self) -> bool {
         matches!(self.node_type(), NodeType::BinaryOperator(_))
     }
+
+    fn find_parent<Callback: Fn(Node) -> bool>(&self, callback: Callback) -> Option<Node> {
+        let mut node = *self;
+
+        while let Some(parent) = node.parent() {
+            if callback(parent) {
+                return Some(parent);
+            }
+            node = parent;
+        }
+
+        None
+    }
 }
 
 pub(crate) fn node_text(node: &Node, contents: &ropey::Rope) -> Option<String> {
     contents.node_slice(node).ok().map(|f| f.to_string())
 }
 
+pub(crate) fn node_has_error(node: &Node) -> bool {
+    // According to the docs, `node.has_error()` should return `true`
+    // if `node` is itself an error, or if it contains any errors, but that
+    // doesn't seem to be the case for terminal ERROR nodes.
+    // https://github.com/tree-sitter/tree-sitter/issues/3623
+    node.is_error() || node.has_error()
+}
+
+pub(crate) fn node_find_string<'a>(node: &'a Node) -> Option<Node<'a>> {
+    if node.is_string() {
+        // Already on a string
+        return Some(*node);
+    }
+    // If we are on one of the following, we return the string parent:
+    // - Anonymous node inside a string, like `"'"`
+    // - `NodeType::StringContent`
+    // - `NodeType::EscapeSequence`
+    node.find_parent(|parent| parent.is_string())
+}
+
```

Comment:
Oh hmm `.ancestors()` not being exclusive is surprising to me, I was not expecting `node` to be an "ancestor" of itself.

The unwrapped call makes more sense since it explicitly takes `first`, making it slightly less surprising when you also get `first` back on the first iteration

```
std::iter::successors(Some(*self), |p| p.parent())
```